### PR TITLE
Feature/json retcon

### PR DESCRIPTION
**Pull request by mdellabitta**

### CODE REVIEW

File: src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    extractStrings(data.get \ "_source" \ "sourceResource" \ "creator")
+      .map(nameOnlyAgent)
+
+  private def extractEdmTimeSpan(fields: List[(String, JValue)]): Option[EdmTimeSpan] = {
+    {
+      val begin = fields.find(_._1 == "begin").map(_._2.toString)
+      val end = fields.find(_._1 == "end").map(_._2.toString)
+      val displayDate = fields.find(_._1 == "displayDate").map(_._2.toString)
+
+      if (begin.nonEmpty && end.nonEmpty && displayDate.nonEmpty) None
```
Comment (moltude) [inline src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala:86]: Shouldn't this be `isEmpty`?

File: src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      .map(nameOnlyConcept)
+
+  override def identifier(data: Document[JValue]): ZeroToMany[String] =
+    extractStrings(data.get \ "_source" \ "sourceResource" \ "identifier")
+
+  override def language(data: Document[JValue]): ZeroToMany[SkosConcept] =
+    extractStrings(data.get \ "_source" \ "sourceResource" \ "language" \ "name")
+      .map(nameOnlyConcept)
+
+  override def place(data: Document[JValue]): ZeroToMany[DplaPlace] =
```
Comment (moltude) [inline src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala:125]: Don't we want to map coordinates where they existing in the sourceResouce data?

File: src/test/resources/json-retcon/washington.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  "_source": {
+    "ingestType": "item",
+    "ingestionSequence": 999999,
+    "ingestDate": "2016-03-29",
+    "@context": "http://dp.la/api/items/context",
+    "aggregatedCHO": "#sourceResource",
+    "@id": "http://dp.la/api/items/a2a415d0f3030e86e0349f90747d1c86",
+    "id": "a2a415d0f3030e86e0349f90747d1c86",
+    "dataProvider": "University of Washington",
+    "originalRecord": "http://ldp.dp.la/ldp/original_record/a2a415d0f3030e86e0349f90747d1c86.xml",
```
Comment (moltude) [inline src/test/resources/json-retcon/washington.json:15]: SMH

File: src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+import org.json4s.JValue
+import org.json4s.JsonAST.{JArray, JObject, JString}
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods.{compact, render}
+
+abstract class JsonRetconMapper extends JsonMapping with JsonExtractor {
+
+  // OreAggregation
+  override def dplaUri(data: Document[JValue]): ZeroToOne[URI] = {
+    data.get \ "_source" \ "@id" match {
```
Comment (moltude) [inline src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala:15]: `@id` is already the full item path. if you want to construct is use `id` or reuse the entire `@id` prop

```  
_source": {
    "@id": "http://dp.la/api/items/8bca6ea2341791bb4b30c471115e76e8",
    "id": "8bca6ea2341791bb4b30c471115e76e8"
}
```
See 
https://raw.githubusercontent.com/dpla/ingestion3/88eca3ec32ba871ec1de2023a16a36addff557dc/src/test/resources/json-retcon/lc.json

File: src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+      case JString(value) => Some(URI(f"http://dp.la/api/items/$value"))
+      case _ => None
+    }
+  }
+
+  override def sidecar(data: Document[JValue]): JValue =
+    ("prehashId", buildProviderBaseId()(data)) ~
+      ("dplaId", extractString(data.get \ "_source" \ "id").getOrElse(""))
```
Comment (moltude) [inline src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala:23]: There should be no `.getOrElse` on this -- we want the extraction on the jsonlRecord() method to fail if the value is not provided and empty string will validate.

**Review by moltude — CHANGES_REQUESTED** [review]
A couple of issues/questions on this

File: src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala (hunk 1 above)
Comment (mdellabitta) [reply src/main/scala/dpla/ingestion3/mappers/providers/JsonRetconMapper.scala:86]: Lolz. Nice catch!

**Review by moltude — APPROVED** [review]

**Pull request merged**